src/object.hpp

```diff
@@ -0,0 +1,447 @@
+    [[nodiscard]] T as() const noexcept
+        requires std::is_integral_v<T> && std::is_signed_v<T>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.intValue);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_integral_v<T> && std::is_unsigned_v<T> && (!std::is_same_v<T, bool>)
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.uintValue);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, double>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return static_cast<T>(obj.f64);
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, std::string> || std::is_same_v<T, std::string_view>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return {obj.stringValue, size()};
+    }
+
+    template <typename T>
+    [[nodiscard]] T as() const noexcept
+        requires std::is_same_v<T, const char *>
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        return obj.stringValue;
+    }
+
+    // Access the underlying value based on the required type or return a default
+    // value otherwise.
+    template <typename T> [[nodiscard]] T as_or_default(T default_value) const noexcept
+    {
+        const auto &obj = static_cast<const Derived *>(this)->ref();
+        if (!is_compatible_type<T>(static_cast<object_type>(obj.type))) {
+            [[unlikely]] return default_value;
+        }
+        return as<T>();
+    }
+
+private:
+    readable_object() = default;
+
+    friend Derived;
+};
+
+// TODO move object view and object converter header
+
+template <typename Derived> class writable_object {
+public:
+    [[nodiscard]] borrowed_object at(std::size_t idx);
+
+    borrowed_object emplace_back(owned_object &&value);
+    borrowed_object emplace(std::string_view key, owned_object &&value);
+
+private:
+    writable_object() = default;
+
+    friend Derived;
+};
+
+// NOLINTNEXTLINE(fuchsia-multiple-inheritance)
+class borrowed_object final : public readable_object<borrowed_object>,
+                              public writable_object<borrowed_object> {
+public:
+    // borrowed_object() = default;
+    explicit borrowed_object(detail::object *obj) : obj_(obj)
+    {
+        if (obj_ == nullptr) {
+            throw std::invalid_argument("null borrowed object");
+        }
+    }
+    explicit borrowed_object(detail::object &obj) : obj_(&obj) {}
+
+    explicit borrowed_object(owned_object &obj);
+
+    [[nodiscard]] detail::object &ref() { return *obj_; }
+    [[nodiscard]] const detail::object &ref() const { return *obj_; }
+    [[nodiscard]] detail::object *ptr() { return obj_; }
+    [[nodiscard]] const detail::object *ptr() const { return obj_; }
```

Comment:
I imagine these are to be made private in the future, right? Same for any other public exposures of default::object.

Reply:
Yes, that's the eventual goal, but I can't as it is.